src/providers/irc/IrcConnection2.cpp

```diff
@@ -13,40 +17,96 @@ namespace {
+    QObject::connect(this, &Communi::IrcConnection::socketError, this,
+                     [this](QAbstractSocket::SocketError error) {
+                         qCDebug(chatterinoIrc) << "Connection error:" << error;
+                     });
+
+    QObject::connect(
+        this, &Communi::IrcConnection::socketStateChanged, this,
+        [this](QAbstractSocket::SocketState state) {
+            if (state == QAbstractSocket::UnconnectedState)
+            {
+                if (!this->expectConnectionLoss_.load())
+                {
+                    // Ensure we always notify our parent that we should
+                    // reconnect; will also be called if we're reconnecting
+                    // manually
+                    this->connectionLost.invoke(false);
+                }
+            }
+        });
+
+    // Schedule a reconnect that won't violate RECONNECT_MIN_INTERVAL
+    this->smartReconnect.connect([this] {
+        if (this->reconnectTimer_.isActive())
+        {
+            return;
+        }
+
+        auto delta =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::steady_clock::now() - this->lastConnected_)
+                .count();
+        delta = delta < RECONNECT_MIN_INTERVAL
+                    ? (RECONNECT_MIN_INTERVAL - delta)
+                    : 10;
+        qCDebug(chatterinoIrc) << "Reconnecting in" << delta << "ms";
+        this->reconnectTimer_.start(delta);
+    });
+
+    this->reconnectTimer_.setInterval(RECONNECT_MIN_INTERVAL);
+    this->reconnectTimer_.setSingleShot(true);
+    QObject::connect(&this->reconnectTimer_, &QTimer::timeout, [this] {
+        if (this->isConnected())
+        {
+            // E.g. user manually reconnecting doesn't cancel this path, so
+            // just ignore
+            qCDebug(chatterinoIrc) << "Reconnect: already reconnected";
+        }
+        else
+        {
+            qCDebug(chatterinoIrc) << "Reconnecting";
+            this->open();
+        }
+    });
+
+    // Send ping every x seconds
     this->pingTimer_.setInterval(5000);
     this->pingTimer_.start();
     QObject::connect(&this->pingTimer_, &QTimer::timeout, [this] {
         if (this->isConnected())
         {
-            if (this->waitingForPong_.load())
+            if (this->recentlyReceivedMessage_.load())
             {
-                // Not sending another ping as we haven't received the matching pong yet
+                // If we're still receiving messages, all is well
+                this->recentlyReceivedMessage_ = false;
+                this->waitingForPong_ = false;
                 return;
             }
 
-            if (!this->recentlyReceivedMessage_.load())
+            if (this->waitingForPong_.load())
+            {
+                // The remote server did not send a PONG fast enough; close the
+                // connection
+                this->expectConnectionLoss_ = true;
```

Comment:
This seems like a noop since we already set `this->expectConnectionLoss_ = true;` in the `IrcConnection::close()` method